Лабораторная работа 1. Тверитнев Михаил

ktveritnev23 opened the pull request:
Вопросы по возможностям современных regex-машин в современных ЯП и конечным автоматам.

ktveritnev23 committed:
Лабораторная работа 1. Тверитнев Михаил

BaldiSlayer commented on data/data.yaml:
```diff
@@ -4649,4 +4649,151 @@
+  4. Поддержка захватывающих (possessive) квантификаторов и атомарных группировок.
+  5. НКА поддерживает обработку многоопределённости.
+  В то же время, ДКА имеет полиномиальное время выполнения, в отличие от экспоненциального для НКА.
+  Комбинирование ДКА и НКА позволяет достигать необходимого баланса между скоростью работы и занимаемым в памяти местом (как известно, НКА с n состояниями может соответствовать ДКА с 2^n состояниями, а НКА имеет в худшем случае экспоненциальное время выполнения). При этом большое количество состояний и объём занимаемой ими памяти также могут негативно влиять на общую производительность (в случае  ...
```
> в случае непопадания в кэш процессорному ядру приходится обращаться в оперативную память, что увеличивает задержки

подушню, но L3 кеш например пошарен между ядрами

BaldiSlayer commented on data/data.yaml:
```diff
@@ -4649,4 +4649,151 @@
+questions:
+- 'Что такое "сильная" LL(k) грамматика и чем она отличается от обычной?
+'
+- answer: 'Обычно переписывание регулярных выражений для их ускорения осуществляется вручную, поскольку общего алгоритма для их оптимизации пока не существует, а доступные способы оптимизации зависят от сценария использования регуляного выражения (например, a{min, max} быстрее (a+)+, но для такой замены необходимо знать, сколько повторяющихся символов может быть во входной строке). Однако есть би ...
```
где-то тут кажется сломан формат yaml

BaldiSlayer replied:
> Регулярка (a+)+ в принципе ужасна )) Почему для её оптимизации нельзя просто заменить её хотя бы на a*(a+) (если вдруг нужна группа захвата) или тупо на a+, если группа захвата не нужна, а понадобились ограничители, мне не очень понятно.

язык (C+)+ 

))

BaldiSlayer commented on data/data.yaml:
```diff
@@ -4649,4 +4649,151 @@
+Современные устройства позволяют обрабатывать данные в несколько потоков (если учитывать возможность вычислений на GPU, то доступны сотни потоков).
+Существует два основных подхода к параллелизации регулярных выражений:
+1. Построение параллельного конечного автомата (англ. parallel finite-state machine, PFSM) для регулярного выражения.
+2. Разбиение данных на несколько частей, каждая из которых обрабатывается движком регулярных выражений в отдельном потоке. Например, в Java для управлении потоками используется библиотека `java.util.concurrent`, в C++ - `std::thread`, а в Go - `goroutines`.
```
TLDR
ну строго говоря горутина != поток) и надо чуть пошаманить, чтобы горутины были на разных логических ядрах

еще подушню, но горутины и `std::thread` это разное с точки зрения того, что это под капотом. `std:thread` является потоком ОС, а вот горутина им не является, горутины шедулятся рантаймом Go, а потоки ОС планировщиком ОС

и тд)
можно долго продолжать, но прям сильно душить уж не буду

скажу только то, что горутина это абстракция над потоком ОС (будет интересно - прочитай про Green threads)

BaldiSlayer commented on data/data.yaml:
```diff
@@ -4649,4 +4649,151 @@
+author: Тверитнев Михаил
+id: 266
+questions:
+  - 'Обратные ссылки с заданной глубиной в интерпретаторе регулярных выражений языка Ruby.
```
а такое ток в Ruby есть?

BaldiSlayer commented on data/data.yaml:
```diff
@@ -4649,4 +4649,151 @@
+- 'Простой способ построения пересечения регулярных языков с помощью lookahead
+  '
+- answer: 'Алгоритм Бржозовского (Brzozowski algorithm) позволяет минимизировать любой конечный автомат, даже недетерминированный, посредством последовательного двукратного применения к нему операций обращения и детерминизации.
+Теорема Бржозовского (англ. Brzozowski theorem, 1962): Пусть A - конечный автомат, распознающий язык L, det(A) - детерминизированный автомат, rev(A) - обратный автомат для А (т.е. такой автомат, в котором начальные и конечные состояния поменяли местами, а направления всех переходов изменены на обратные). Минимальный ДКА для автомата А может быть построен следующим образом: Amin = det(rev(det(rev( ...
```
1962 нам если честно не очень важно

BaldiSlayer commented on data/data.yaml:
```diff
@@ -4649,4 +4649,151 @@
+S -> aAa
+S -> bA2ba
+A -> b | e
+A2 -> b | e.
```
давай A_2 заменим условно на C ) не вижу смысла юзать нижний регистр там, где можно обойтись без него

TonitaN commented on data/data.yaml:
```diff
@@ -4649,4 +4649,151 @@
+'
+Современные устройства позволяют обрабатывать данные в несколько потоков (если учитывать возможность вычислений на GPU, то доступны сотни потоков).
+Существует два основных подхода к параллелизации регулярных выражений:
+1. Построение параллельного конечного автомата (англ. parallel finite-state machine, PFSM) для регулярного выражения.
```
На этот случай в регулярки вводится замечательная операция асинхронной композиции или шаффла, #. Её замечательность - в том, что она коммутативна, и алгебра регулярок с такой операцией позволяет определить производную произведения (асинхронной композиции) так же, как она определяется в классическом матанализе. Рекомендую в этом ответе добавить определение шаффла, раз уж пошла такая история.

TonitaN commented on data/data.yaml:
```diff
@@ -4649,4 +4649,151 @@
+'
+- answer: 'В PCRE2-совместимых регулярных выражениях (полное название стандарта - Perl Compatible Regular Expressions) предусмотрено использование нескольких способов обращения к скобочным группам, или группам захвата (англ. captive group):
+1. Прямое обращение по номеру. Группы в регулярках нумеруются слева направо, начиная с 1.
+2. Именованные группы. Для задания имени группе захвата используется следующий синтаксис: `(?<name>)`, где <name> - имя группы. Примеры:
```
Тут не хватает примеров

TonitaN commented on data/data.yaml:
```diff
@@ -4649,4 +4649,151 @@
+    Пример: `^(?<Counter>A)+ (?<-Counter>B)+ (?(Counter)(?!))$` распознаёт слова из языка L = {A^nB^n | n ∈ N, n > 0 }.
+    Более того, со стека можно снимать несколько захваченных символов. Например, регулярное выражение `^(?=.*[a-z](?<N>)|)(?=.*[A-Z](?<N>)|)(?=.*[0-9](?<N>)|)(?<-N>){2}.{8,}` распознаёт слова (пароли), состоящие не менее чем из восьми символов и содержащие не менее двух из трёх видов символов: строчные буквы, прописные буквы, цифры.
+'
+author: Тверитнев Михаил
```
Вопрос отрезался.

TonitaN reviewed (commented):
Материал интересный и нетривиальный, хотя иногда очень специфический, но понравилось.

TonitaN commented:
На всякий случай, чтобы не было "никаких разночтений (с)" - когда я предлагаю правки, их совсем не обязательно и обычно не желательно вставлять в ответ методом copy&paste (исключение - если речь о точных формулировках). Всегда лучше подстраивать предложение под контекст. Да и некрасиво, если стилистика ответа начинает резко меняться, потому что часть писал студент, а часть - получается что я. Не к тому, что вы так делали, а к тому, что так делали некоторые другие, и на этот образец равняться не надо.

Тегну здесь ещё @ark2016 с аналогичным пожеланием, чтобы вам не казалось, что про вас забыли)

ktveritnev23 committed:
Update data.yaml

1) Добавлен обрезанный вопрос с id=274;
2) Добавлены примеры для вопроса с id=273

TonitaN commented:
@ktveritnev23 , по всем правкам нужно пройтись, и не в последнюю ночь перед комиссией
